feat(datasource/endoflife-date): add endoflife.date datasource (#21994)
renovatebot/renovate: lib/modules/datasource/endoflife-date/readme.md

[endoflife.date](https://endoflife.date) provides version and end-of-life information for different packages.

> To find the appropriate "package" name for the software you're trying to update, use the endoflife.date "All packages" API endpoint. You can find it in [the API documentation](https://endoflife.date/docs/api).

Reviewer: ```suggestion
To find the appropriate "package" name for the software you're trying to update, use the endoflife.date "All packages" API endpoint.
You can find it in [the endoflife.date API documentation](https://endoflife.date/docs/api).
```

> This datasource uses `loose` versioning by default. If the software you are upgrading supports a versioning that is more strict, e.g. `semver`, it is recommended to specify that explicitly.

Reviewer: ```suggestion
By default, this datasource uses `loose` versioning.
If possible, we recommend you use a stricter versioning like `semver` instead of `loose.
```

**Usage Example**

> Imagine using Amazon EKS and wanting to update the versions in a terraform `.tfvars` file.

Reviewer: ```suggestion
Say you're using Amazon EKS and want Renovate to update the versions in a Terraform `.tfvars` file.
```

> 
> An example `.tfvars` file would look as follows:

Reviewer: ```suggestion
For example, you have this `.tfvars` file:
```

```hcl
# renovate: datasource=endoflife-date depName=amazon-eks versioning=loose
kubernetes_version = "1.26"
```

> then, add the following configuration to your `renovate.json`:

Reviewer: ```suggestion
Give the above `.tfvars` file, you put this in your `renovate.json`:
```

```json
{

>   "regexManagers": [

Reviewer: Once the config is correct, do we want to add it to the [Preset Regex Managers](https://docs.renovatebot.com/presets-regexManagers/) page?

Code for the presets lives here:

- https://github.com/renovatebot/renovate/blob/main/lib/config/presets/internal/regex-managers.ts
- https://github.com/renovatebot/renovate/blob/main/lib/config/presets/internal/regex-managers.spec.ts

Author: Having this config in the presets makes sense, yes. It doesn't have anything to do with this PR per se, so I think it makes sense to add it in a separate PR where it can be discussed independently. 
I can take care of that PR, too.

What do you think?

    {

>       "fileMatch": ["^Dockerfile$"],

Reviewer: This `fileMatch` is pointing at a Dockerfile. But the Terraform config file extension in the examples is `.tfvars`. This looks wrong to me. 😄

[endoflife.date](https://endoflife.date) provides version and end-of-life information for different packages.

To find the appropriate "package" name for the software you're trying to update, use the endoflife.date "All packages" API endpoint.
You can find it in [the endoflife.date API documentation](https://endoflife.date/docs/api).

By default, this datasource uses `loose` versioning.
If possible, we recommend you use a stricter versioning like `semver` instead of `loose`.

**Usage Example**

Say you're using Amazon EKS and want Renovate to update the versions in a Terraform `.tfvars` file.
For example, you have this `.tfvars` file:

```hcl
# renovate: datasource=endoflife-date depName=amazon-eks versioning=loose
kubernetes_version = "1.26"
```

> Give the above `.tfvars` file, you put this in your `renovate.json`:

Reviewer: ```suggestion
Given the above `.tfvars` file, you put this in your `renovate.json`:
```

Whoops I made a typo here. 😄

```json
{
  "regexManagers": [
    {
      "fileMatch": [".+\\.tfvars$"],
      "matchStrings": [
        "#\\s*renovate:\\s*datasource=(?<datasource>.*?) depName=(?<depName>.*?)( versioning=(?<versioning>.*?))?\\sENV .*?_VERSION=\"(?<currentValue>.*)\"\\s"
      ],
      "versioningTemplate": "{{#if versioning}}{{{versioning}}}{{/if}}"
    }
  ]
}

> ```

Reviewer: How about adding a line below the config itself that explains what it does?

Author: Addressed in 5ae5709ed0c9036c70a3a975a204db6bee4c1054. I also fixed the regexManager configuration since it was broken. Thanks for pointing me at it!

You can find it in [the endoflife.date API documentation](https://endoflife.date/docs/api).

By default, this datasource uses `loose` versioning.
If possible, we recommend you use a stricter versioning like `semver` instead of `loose`.

**Usage Example**

Say you're using Amazon EKS and want Renovate to update the versions in a Terraform `.tfvars` file.
For example, you have this `.tfvars` file:

```hcl
# renovate: datasource=endoflife-date depName=amazon-eks versioning=loose
kubernetes_version = "1.26"
```

Given the above `.tfvars` file, you put this in your `renovate.json`:

```json
{
  "regexManagers": [
    {
      "description": "Update Kubernetes version for Amazon EKS in tfvars files",
      "fileMatch": [".+\\.tfvars$"],
      "matchStrings": [
        "#\\s*renovate:\\s*datasource=(?<datasource>.*?) depName=(?<depName>.*?)( versioning=(?<versioning>.*?))?\\s.*?_version\\s*=\\s*\"(?<currentValue>.*)\""
      ],
      "versioningTemplate": "{{#if versioning}}{{{versioning}}}{{/if}}"
    }
  ],
  "packageRules": [
    {
      "matchDatasources": ["endoflife-date"],
      "matchPackageNames": ["amazon-eks"],
      "extractVersion": "^(?<version>.*)-eks.+$"
    }
  ]
}
```

> With this configuration, renovate will parse all `*.tfvars` files in the repository. It will then update variables that end with `_version` and have the `# renovate: datasource=endoflife-date depName=dependency-name versioning=versioning` comment set in the line above when any new versions are available.

Reviewer: ```suggestion
With this configuration, Renovate will parse all `*.tfvars` files in the repository.
It will then update variables that end with `_version` and have the `# renovate: datasource=endoflife-date depName=dependency-name versioning=versioning` comment set in the line above when any new versions are available.
```
